Sources/SourceKit/SourceKitServer.swift

```diff
@@ -256,29 +357,68 @@ extension SourceKitServer: BuildSystemDelegate {
     return documentManager.openDocuments.intersection(changes)
   }
 
-  public func fileBuildSettingsChanged(_ changedFiles: Set<DocumentURI>) {
+  /// Handle a build settings change notification from the `BuildSystem`.
+  /// This has two primary cases:
+  /// - Initial settings reported for a given file, now we can fully open it
+  /// - Changed settings for an already open file
+  public func fileBuildSettingsChanged(
+    _ changedFiles: [DocumentURI: FileBuildSettingsChange]
+  ) {
     queue.async {
       guard let workspace = self.workspace else {
         return
       }
       let documentManager = workspace.documentManager
-      for uri in self.affectedOpenDocumentsForChangeSet(changedFiles, documentManager) {
+      let openDocuments = documentManager.openDocuments
+      for (uri, change) in changedFiles {
+        // Non-ready documents should be considered open even though we haven't
+        // opened it with the language service yet.
+        guard openDocuments.contains(uri) else { continue }
+        guard self.documentsReady.contains(uri) else {
+          // Case 1: initial settings for a given file. Now we can process our backlog.
+          log("Initial build settings received for opened file \(uri)")
+
+          guard let snapshot = documentManager.latestSnapshot(uri),
+                let service = self.languageService(for: uri, snapshot.document.language, in: workspace) else {
+            log("No snapshot or language service for opened file \(uri)", level: .error)
```

Comment:
Hmm, interesting. This case shouldn't happen, if it does we have a file that's considered open which the build system knows about yet we don't have a language service for I guess it's possible if the LS crashes? Cancelling makes sense IMO but once we're in this state we won't get out of it, although only documentedits will queue up, maybe we should change the behavior of changeDocument to account for this? WDYT?

Reply:
We know it had a language service when we opened it (and indeed a reference to that service is captured in the operation), but it's possible that configuration has changed in the meantime - e.g. available toolchains have changed.  In fact, another failure mode here would be that the language service is available, but different than the one we had before.  We could store the language service explicitly with the queue so that we can be sure to get the same one here and never fail.

Reply:
From what I can tell we don't support toolchain for a file changing - in fact once we create the toolchain for a file we can't overwrite it, we always use the cached version.

I think only `openDocument` should use the `self.languageService(for: uri, snapshot.document.language, in: workspace)` call, but this method here should always use the cached version and never create a new one (in practice this is what happens at the moment).

Reply:
And hmm storing it seems like a good idea, but if we never have a ToolchainLanguageServer for a file, how can we queue up the edit requests properly? We probably shouldn't right? The code here at the moment does...

Reply:
> From what I can tell we don't support toolchain for a file changing - in fact once we create the toolchain for a file we can't overwrite it, we always use the cached version.

Fair enough - we ought to support that someday, but it's fine for now.

Since we think this situation should never happen and it should always get the cached one, how about logging an error and cancelling all the requests?  That will make it obvious that something is going wrong if this ever changes.

Reply:
Done